### ISSUE TITLE
Floats broken in PyYAML

### ISSUE THREAD

**Issue by LourensVeen** [yatiml/yatiml#53]
Floating point values aren't parsed correctly in PyYAML: https://stackoverflow.com/questions/30458977/yaml-loads-5e-6-as-string-and-not-a-number

So now that we're using that, YAtiML is also broken. Can we fix this in YAtiML, maybe add a check that, if something gets parsed as a string, tries whether is isn't actually a float using the correct regex?

**Comment by LourensVeen** [comment]
It seems that what YAML considers a float depends on the version of YAML you're using. PyYAML is still on YAML 1.1, while ruamel.yaml has been updated to YAML 1.2 (Anthon is wrong in the link above about PyYAML being incorrect, it's just outdated). Despite the minor version difference, there are noticeable differences in how literals are parsed.

The YAML 1.1 spec has [this regex for floats](https://yaml.org/type/float.html):
```
 [-+]?([0-9][0-9_]*)?\.[0-9.]*([eE][-+][0-9]+)? (base 10)
|[-+]?[0-9][0-9_]*(:[0-5]?[0-9])+\.[0-9_]* (base 60)
|[-+]?\.(inf|Inf|INF) # (infinity)
|\.(nan|NaN|NAN) # (not a number)
```
where [YAML 1.2.2 says](https://yaml.org/spec/1.2.2/#tag-resolution):
```
[-+]? ( \. [0-9]+ | [0-9]+ ( \. [0-9]* )? ) ( [eE] [-+]? [0-9]+ )?
[-+]? ( \.inf | \.Inf | \.INF )
\.nan | \.NaN | \.NAN
```

This explains why `1e-12` in the ymmsl-python test didn't parse as a float anymore, because it's valid YAML 1.2 but invalid YAML 1.1.

So with the switch from ruamel.yaml to PyYAML we have switched from YAML 1.2 to YAML 1.1 and now what we accept is different. That's not so good. Then again, YAML is a mess anyway and so are PyYAML and ruamel.yaml, so we should probably just be pragmatic and make YAtiML accept the things that people would generally expect to be a float value while leaving things that people would not expect to be a float value to be a string, if we can.

So what's the difference?

YAML 1.1:
- allows underscores in the integer part
- allows extra periods in the decimal part
- requires a decimal point at all times

YAML 1.2.2:
- doesn't allow underscores in the integer part
- doesn't allow extra periods in the decimal part
- doesn't require a decimal point, and also matches integers (but those have already been matched presumably by the time we get to trying to match a float)

(let's not talk about that base-60 float in YAML 1.1...).

So it seems like it's best if we just adopt the YAML 1.2.2 version, that also matches what the rest of the world does.

PyYAML implements this stuff in the `Resolver`. We don't have a custom resolver, but with PyYAML's horrible mixin/inheritance-based design we should be able to mix something in somewhere. `SafeLoader` inherits from `Resolver`, so it's going to have the `yaml_implicit_resolvers` class variable that contains the YAML 1.1 float resolver. So we could patch that in our `Loader` to parse YAML 1.2 floats instead.

Alternatively, we can let PyYAML parse first and then patch it up in the recognition stage. For that, we'll want to check everything tagged as a float and re-parse it, possibly turning it back into a string, and everything tagged as a string and possibly turn it into a float. Or actually, we could add a Patcher that gets called before the Recognizer that patches things up? That's cleaner but still messy.

**Comment by LourensVeen** [comment]
Patched it in the loader. If we need more patches we can split them off.

**Comment by LourensVeen** [comment]
Released with 0.11.1.
